src/System.Private.CoreLib/shared/System/Collections/ObjectModel/Collection.cs

```diff
@@ -162,6 +170,57 @@ protected virtual void SetItem(int index, T item)
             items[index] = item;
         }
 
+        protected virtual void InsertItemsRange(int index, IEnumerable<T> collection)
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
+
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            int i = 0;
+            foreach (T item in collection)
+            {
+                Insert(index + i, item);
+                i++;
+            }
+        }
+
+        protected virtual void RemoveItemsRange(int index, int count)
+        {
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            int indexCount = index + count;
+            int length = indexCount > items.Count ? items.Count : indexCount;
```

Comment:
I still think there is a concern if `index + count` > int.MaxValue. Currently if that's true, length will be < 0, which means we will not remove anything. Is that the expected behavior?

Reply:
The goal of this code is to get the number of items to iterate through starting at the position of index. We need to constrain the count to the maximum length of the array otherwise we will get an `IndexOutOfRangeException`. There are also cases where we won't reach the end of the array.

If `index + count` > `int.MaxValue` is true, maybe we should constrain to `items.Count`.

Here is my proposal to clean this up:

```c#
int indexCount = index + count;
int length = indexCount < 0 || indexCount > items.Count ? items.Count : indexCount;
```

@ahsonkhan what are your thoughts on this?

Reply:
```C#
int length = items.Count;

// equivalent to (index + count < items.Count) without possibility of overflow
if (index < items.Count - count)
{
   length = index + count;
}
```

Reply:
This makes sense to me. I'll update the code to reflect this proposal

Reply:
I added additional unit tests and the overflow exception was happening, this is a great catch that I did not think of when writing the implementation

Reply:
> We can avoid the extra checks if we re-write it this way

FYI, we can't always re-write it like this since it can result in underflow (lower than `int.MinValue`), but in this case `items.Count >= 0` so `items.Count - count` can never be less than `int.MinValue`.